### ISSUE TITLE
wrong patch(field not in schema) can leading to unintentional deletion of managed resource

### ISSUE THREAD

**Issue by haarchri** [crossplane-contrib/function-patch-and-transform#59]
I've come across a peculiar issue related to the patch and transform function. Here's the scenario:

- Deployed an XR successfully.
- All managed resources are in a ready state.
- Attempted to add a new patch, but encountered an issue.
- The problem arises when the field I intend to patch is not present in the schema.
- As a consequence, the operation results in the deletion of the managed resource.

```
Events:
  Type    Reason                   Age    From                                                                       Message
  ----    ------                   ----   ----                                                                       -------
  Normal  DeletedExternalResource  4m49s  managed/containerservice.azure.upbound.io/v1beta1, kind=kubernetescluster  Successfully requested deletion of external resource

```
Steps to Reproduce:

- Deploy an XR.
- Ensure all managed resources are ready.
- Attempt to apply a patch to a field not present in the schema.

**Comment by phisco** [comment]
Interesting. Can you provide some more details about the patch? Any other relevant log/event? What's the result of runnig the same composition using `crossplane beta render`?

**Comment by JonathanO** [comment]
I think this issue might have been due to a type error where a field being rendered into is the wrong type for the patch being applied, rather than a schema validation failure.

Attempting to patch into e.g. spec.example[0] when spec.example has been initialised in the state as a non-array type already (including nil I think) leads to a "spec.example is not an array" error being generated from here:
https://github.com/crossplane/crossplane-runtime/blob/d23a82b3a2f5ac662c148171e716bb18e208f82e/pkg/fieldpath/paved.go#L359-L377
The function fails to render the resource, but only returns a Warning in the result so Crossplane treats it as if the resource is being removed.

I wonder if this (and similar errors which wont resolve themselves by continuing to attempt to apply the patches) should be Fatal rather than Warning?

**Comment by haarchri** [comment]
```
kubectl get managed
NAME                                                  READY   SYNCED   EXTERNAL-NAME                          AGE
resourcegroup.azure.upbound.io/test-aks-fhkhd-6wqx8   True    True     b9925bc4-8383-4c37-b9d2-fa456d1bb1c7   6m32s

NAME                                                   KIND     PROVIDERCONFIG   SYNCED   READY   AGE
object.kubernetes.crossplane.io/test-aks-fhkhd-jw966   Secret   default          True     True    2m7s
```

add a patch with wrong `toFieldPath`

```
              - type: FromCompositeFieldPath
                fromFieldPath: spec.parameters.region
                toFieldPath: spec.forProvider.location[0].region # <-- location is not an array
```


function output:
```
{"level":"info","ts":1703412840.0993283,"caller":"fn/fn.go:35","msg":"Running Function","tag":""}
{"level":"info","ts":1703412840.099947,"caller":"fn/fn.go:238","msg":"Cannot render FromComposite patches for composed resource","tag":"","xr-version":"platform.haarchri.io/v1alpha1","xr-kind":"XCluster","xr-name":"test-aks-fhkhd","resource-template-name":"resource-group","warning":"cannot apply the \"FromCompositeFieldPath\" patch at index 2: cannot patch to object: spec.forProvider.location is not an array"}
{"level":"info","ts":1703412840.100242,"caller":"fn/fn.go:270","msg":"Successfully processed patch-and-transform resources","tag":"","xr-version":"platform.haarchri.io/v1alpha1","xr-kind":"XCluster","xr-name":"test-aks-fhkhd","resource-templates":2,"existing-resources":2,"warnings":1}
```

after the function run - the managed resource resourcegroup.azure.upbound.io/test-aks-fhkhd-6wqx8 gets deleted:

```
kubectl get managed
NAME                                                  READY   SYNCED   EXTERNAL-NAME                          AGE
resourcegroup.azure.upbound.io/test-aks-fhkhd-6wqx8   False   True     b9925bc4-8383-4c37-b9d2-fa456d1bb1c7   6m58s

NAME                                                   KIND     PROVIDERCONFIG   SYNCED   READY   AGE
object.kubernetes.crossplane.io/test-aks-fhkhd-jw966   Secret   default          True     True    2m33s
```

```
  Normal   DeletedExternalResource      20s                 managed/azure.upbound.io/v1beta1, kind=resourcegroup  Successfully requested deletion of external resource
```

```
kubectl get managed
NAME                                                   KIND     PROVIDERCONFIG   SYNCED   READY   AGE
object.kubernetes.crossplane.io/test-aks-fhkhd-jw966   Secret   default          True     True    6m18s
```

**Comment by negz** [comment]
https://github.com/crossplane-contrib/function-patch-and-transform/blob/cf7a6f78/render.go#L130
https://github.com/crossplane-contrib/function-patch-and-transform/blob/cf7a6f78/fn.go#L202

I think the issue is that I made an incorrect assumption here (note the code was refactored a bit since the version you're running). The code assumes that if we encounter an error while applying a patch it must be because a required field path was not found. It therefore thinks that because the patch is required, it should skip returning the resource from desired state altogether. Crossplane interprets the resource no longer appearing in the desired state as the function asking for it to be deleted.

Obviously this assumption is wrong and we could hit an error for a number of reasons, including a misconfigured patch.

In c/c when we hit any error patching a composed resource we just skip updating that resource and move on:

https://github.com/crossplane/crossplane/blob/82e0f1/internal/controller/apiextensions/composite/composition_pt.go#L265

We don't really have that option in this function, because we need to return our desired state every time we're called. We can't just return the observed composed resource, because that will include a bunch of extra fields that aren't really part of our desired state.

I think probably the best thing to do would just be to always return whatever desired state we accumulated, even if we encountered an error in a patch. There's a couple of potential downsides to this approach, but I think it's the best option we have.

**Comment by JonathanO** [comment]
> I think probably the best thing to do would just be to always return whatever desired state we accumulated, even if we encountered an error in a patch.

What about returning a SEVERITY_FATAL Result for errors other than required fields not found? That seems fair since the error is likely caused by a mistake in the composition, and I think it avoids applying partially rendered desired state (https://github.com/upbound/crossplane/blob/b9ed0d3fd9481af51ed70410e19336f7aff91a9e/cmd/crank/beta/render/render.go#L182-L183)

**Comment by negz** [comment]
> What about returning a SEVERITY_FATAL Result for errors other than required fields not found?

Maybe. 🤔

One thing we need to be careful about is patches that will be eventually successful. i.e. Patch N will only succeed once patch N+1 has succeeded. If we returned SEVERITY_FATAL on patch N, patch N+1 could never run to unblock patch N.

It could be that's not a problem though. We already squash errors that occur due to an optional field path not existing. If that's the only scenario that would block another patch we might SEVERITY_FATAL might be fine.

**Comment by phisco** [comment]
> We don't really have that option in this function, because we need to return our desired state every time we're called. We can't just return the observed composed resource, because that will include a bunch of extra fields that aren't really part of our desired state.
> 
> I think probably the best thing to do would just be to always return whatever desired state we accumulated, even if we encountered an error in a patch. There's a couple of potential downsides to this approach, but I think it's the best option we have.

I agree, that would mean we return incomplete resources which then either fail to apply or, even worst, are successfully applied with some default value instead of the ones we wanted, so to me it feels like we miss something on Crossplane side rather than here.

We have 3 (+`UNKNOWN`) severities for `Results`:`FATAL` means everything is broken and the reconciliation should not continue at all, `WARNING` only emits logs and events, `NORMAL` means everything is fine. But there is no way for functions to say that some resource should be ignored. If we had some way to specify that, this function could mark resources with errors during patches as such and Crossplane could just skip them, avoiding deleting them as happens here and in https://github.com/crossplane-contrib/function-patch-and-transform/issues/70.

`Results` don't allow specifying what resource originated the issue, so we could think of adding this information and change the behavior of `FATAL` or `WARNING` severity results, or just add to the `Resource` struct a field to mark the resource as to be ignored. I think the latter is a better option, WDYT?

**Comment by JonathanO** [comment]
> If we had some way to specify that, this function could mark resources with errors during patches as such and Crossplane could just skip them

Agreed, I'm not opposed to fixing both #59 and #70 in the same way. Although I still have a feeling that it might be safer to treat errors caused by bugs in the input or function (#59) as fatal. It might also be beneficial to close this issue, while working on the more complex skipping fix for #70.

> Results don't allow specifying what resource originated the issue, so we could think of adding this information and change the behavior of FATAL or WARNING severity results, or just add to the Resource struct a field to mark the resource as to be ignored. I think the latter is a better option, WDYT?

I think the former would warrant adding a new severity (ERROR?), rather than repurposing FATAL or WARNING.

What should happen with the result of the other patches for the resource that encountered the error? Should they still be attempted? Could that lead to potentially patching the Composite or Environment with inconsistent information? If this was implemented as a flag on Resource then might further pipeline steps read the incomplete desired state? I suspect the behaviour we'd want would depend on the type of error that was encountered (which is why I came back to wondering if FATAL is the safe solution for most errors.)

**Comment by negz** [comment]
I'm leaning toward using `SEVERITY_FATAL`.

I just quickly skimmed the errors that [`RenderComposedPatches`](https://github.com/crossplane-contrib/function-patch-and-transform/blob/11be501f8e04cd1684cce0b3539438f0b6facd20/fn.go#L195) can return to refresh my memory. The errors all seem to fit into one of three buckets:

1. The patch or one of its transforms is misconfigured.
2. The patch is from a required field path that is not yet found.
3. The object being patched to or from can't be serialized to or from JSON.

For misconfigurations the simplest thing to do is to return `SEVERITY_FATAL`. For a new XR this would mean no composed resources would ever be created. For an existing XR this means the desired state of _all resources_ would effectively be frozen until the misconfiguration is fixed. I like that it's simple and easy to reason about, and I don't think we lose much. I'm not sure there's much benefit to applying partial state when we know human intervention is needed to apply it all.

Bucket 2 is tracked in #70 so I'll weigh in on that over there.

I think we can ignore bucket 3 - it's very unlikely to ever happen.
